react/src/component/util.js

```diff
@@ -0,0 +1,81 @@
+/* eslint eqeqeq: 0 */
+/* eslint no-unused-vars: 0 */
+import { useEffect, useLayoutEffect, useRef, useState } from 'react';
+
+export function constrainZoom(transformMatrix, size, dimension) {
+    if (transformMatrix[`scale${dimension}`] < 1) {
+        transformMatrix[`scale${dimension}`] = 1;
+        transformMatrix[`translate${dimension}`] = 0;
+    }
+    else {
+        if (transformMatrix[`translate${dimension}`] > 0) {
+            transformMatrix[`translate${dimension}`] = 0;
+        }
+        else if (size * transformMatrix[`scale${dimension}`] + transformMatrix[`translate${dimension}`] < size) {
+            transformMatrix[`translate${dimension}`] = size - size * transformMatrix[`scale${dimension}`];
+        }
+    }
+    return transformMatrix;
+}
+
+//TODO(pjm): use a library?
+export function debounce(fn, ms) {
+    let timer
+    return _ => {
+        clearTimeout(timer)
+        timer = setTimeout(_ => {
+            timer = null
+            fn.apply(this, arguments)
+        }, ms)
+    };
+}
+
+const xAxisSize = 30;
+const yAxisSize = 60;
+const margin = 25;
+
+function heightAdjustment() {
+    return xAxisSize + margin * 2;
+}
+
+function widthAdjustment() {
+    return yAxisSize + margin * 2;
+}
+
+export function graphMetrics(dim) {
+    return {
+        graphHeight: dim.height - heightAdjustment(),
+        graphWidth: dim.width - widthAdjustment(),
+        graphX: yAxisSize + margin,
+        graphY: margin,
+    };
+}
+
+export function useRefSize(ref, aspectRatio) {
```

Comment:
```suggestion
export function useRefSize(ref) {
```

Unless it is not possible to do this later, this function should not do aspect ratio scaling and other size adjustments. It is called 'useRefSize' so it should only return the size of a ref. This is more reusable and the other scaling should be made into other functions

Reply:
I've made useRefSize() reusable - it now only returns the dimensions of the component ref. I added useGraphContentBounds() which uses useRefSize() and sets the component height based on the graph's aspect ratio, and returns the graph position info.

I also updated the method names to match your suggestions, but take a new look at utility/component.js to make sure it looks OK.